lamindb/core/history/_trigger_installer.py

```diff
@@ -0,0 +1,229 @@
+import enum
+import re
+from abc import ABC, abstractmethod
+
+from django.apps import apps
+from django.db.backends.base.base import BaseDatabaseWrapper
+from django.db.backends.utils import CursorWrapper
+from lamin_utils import logger
+
+from lamindb.models.history import HistoryLock
+
+
+class HistoryEventTypes(enum.Enum):
+    INSERT = 0
+    UPDATE = 1
+    DELETE = 2
+
+
+# Certain tables must be excluded from history triggers to avoid
+# creating infinite loops of triggering.
+EXCLUDED_TABLES = ["lamindb_history", "lamindb_historylock"]
+
+
+class HistoryRecordingTriggerInstaller(ABC):
+    """Installs triggers that record history into the database's tables."""
+
+    def __init__(self, connection: BaseDatabaseWrapper):
+        self.connection = connection
+
+    @abstractmethod
+    def install_triggers(self, table: str):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_tables_with_installed_triggers(self) -> set[str]:
+        raise NotImplementedError()
+
+    def update_history_triggers(self):
+        # Ensure that the history lock exists
+        HistoryLock.load()
+
+        tables = self._get_db_tables()
+
+        tables_with_installed_triggers = self.get_tables_with_installed_triggers()
+
+        tables_missing_triggers = tables.difference(tables_with_installed_triggers)
+
+        for table in tables_missing_triggers:
+            if table not in EXCLUDED_TABLES:
+                logger.info(f"Installing history recording triggers for {table}")
+                self.install_triggers(table)
+
+    def _get_db_tables(self) -> set[str]:
+        """Get the schema and table names for all tables in the database for which we might want to track history."""
+        all_models = apps.get_models()
+
+        tables = set()
+
+        for model in all_models:
+            # Record the model's underlying table, as well as
+            # the underlying table for all of its many-to-many
+            # relationships
```

Comment:
I think it helps here that Django doesn't define ManyToMany symmetrically the way SQLAlchemy does. Hence just iterating over `_meta.many_to_many` seems indeed a simple and reliable solution. ☺️

Reply:
Yeah, that was a pleasant surprise.